scala/runtime/src/test/scala/org/coursera/courier/templates/EnumTemplateRaceTest.scala

```diff
@@ -0,0 +1,201 @@
+    private val foreignObjectA = clazzA.getConstructor().newInstance()
+    private val foreignObjectB = clazzB.getConstructor().newInstance()
+    private val foreignObjects = List(foreignObjectA, foreignObjectB)
+
+
+    def value(i:Int):String = s"""${method("value").invoke(foreignObjects(i))}"""
+    def withName(i:Int):String = s"""${method("withName").invoke(foreignObjects(i))}"""
+
+    def race( a : => Unit, b : => Unit, useStartSignal:(Int,Int)=(-1,-1)):Boolean = {
+      logging.set(true)
+      val aThread = new Thread("A") {
+        waitForStartSignal.set(useStartSignal._1)
+        override def run() {
+          a
+        }
+      }
+      val bThread = new Thread("B") {
+        waitForStartSignal.set(useStartSignal._2)
+        override def run() {
+          b
+        }
+      }
+      aThread.start()
+      bThread.start()
+      Thread.sleep(1)
+      startSignal.countDown()
+      aThread.join(10000)
+      bThread.join(10000)
+      val aAlive = aThread.isAlive
+      val bAlive = bThread.isAlive
+      if (aAlive) aThread.interrupt()
+      if (bAlive) bThread.interrupt()
+      ! (aAlive || bAlive)
+    }
+  }
+  @Test
+  def raceValueValue():Unit = {
+    val testSetup = new TestSetup()
+    testSetup.value(0)
+    testSetup.withName(0)
+    assert(testSetup.race(testSetup.value(1), testSetup.value(1)))
+  }
+  @Test
+  def raceValueWithName():Unit = {
+    val testSetup = new TestSetup()
+    testSetup.value(0)
+    testSetup.withName(0)
+    assert(testSetup.race(testSetup.value(1), testSetup.withName(1)))
+  }
+  @Test
+  def raceWithNameWithName():Unit = {
+    val testSetup = new TestSetup()
+    testSetup.value(0)
+    testSetup.withName(0)
+    assert(testSetup.race(testSetup.withName(1), testSetup.withName(1)))
+  }
+
+}
+
+object EnumTemplateRaceTest {
+  // This test is for a heisenbug. If it recurs, please increase this value.
+  val NUMBER_OF_RUNS = 1
+  @Parameters(name = "{0}")
+  def parameters: java.util.Collection[Array[AnyRef]] = {
+    val testCases = new java.util.ArrayList[Array[AnyRef]]
+    (1 to NUMBER_OF_RUNS).foreach { i =>
+      testCases.add(Array[AnyRef](i.toString))
+    }
+    testCases
+  }
+}
+
+/* The remaining test classes and objects resemble those created
+   by Courier.
+ */
+sealed abstract class EnumTypeA(name: String, properties: Option[DataMap])
+  extends ScalaEnumTemplateSymbol(name, properties) {
+}
+
+object EnumTypeA extends ScalaEnumTemplate[EnumTypeA] {
+  case object VALUE1 extends EnumTypeA("VALUE1", properties("VALUE1"))
+  case object VALUE2 extends EnumTypeA("VALUE2", properties("VALUE2"))
+  val SCHEMA = DataTemplateUtil.parseSchema(
+    """
+      { "type": "enum",
+        "name": "EnumTypeA",
+        "namespace": "org.coursera.courier.templates",
+        "symbols": ["VALUE1", "VALUE2"]
+      }""").asInstanceOf[EnumDataSchema]
+  override def withName(s: String): EnumTypeA = {
+    symbols.find(_.toString == s).get
```

Comment:
Note this depends on [findSymbols](https://github.com/coursera/courscala/blob/master/courscala/src/main/scala/org/coursera/common/collection/Enum.scala#L63) which may be the code that might actually need 'fixing' ...